### ISSUE TITLE
increased GPU usage in 1.31.1

### ISSUE THREAD

**Issue by unic0rn** [microsoft/vscode#69270]
in 1.31.1 my GPU (IGP in AMD A8-7600) stays at 655MHz all the time. 1.30.2 (the previous version i've used), opened in the same exact state, lets GPU sit at 351MHz.

this may definitely decrease battery life in case of laptops. in case of desktops, that's still a bug.

**Comment by bpasero** [comment]
We are building exploration builds that use a much newer version of our UI framework (Electron version 4.0.x). I wonder if this issue reproduces with one of these builds, could you try? Download:
* macOS: [Download](https://az764295.vo.msecnd.net/exploration/14236f9f44cc7c94055bb8e65ebda29a4a91ddc7/VSCode-darwin-exploration.zip)
* Linux: [Download](https://az764295.vo.msecnd.net/exploration/14236f9f44cc7c94055bb8e65ebda29a4a91ddc7/code-exploration_1.32.0-1550490730_amd64.deb)
* Windows: [Download](https://az764295.vo.msecnd.net/exploration/14236f9f44cc7c94055bb8e65ebda29a4a91ddc7/VSCodeUserSetup-x64-1.32.0-exploration.exe)

**Comment by unic0rn** [comment]
not to be picky, but is there a .zip version? because that's what i'm using, with manual update.

**Comment by bpasero** [comment]
@unic0rn https://az764295.vo.msecnd.net/exploration/6e3c15ae7b1abcd62201218720ccd996a38adad3/VSCode-win32-x64-1.32.0-exploration.zip

**Comment by unic0rn** [comment]
thanks, will test in a moment and let you know.

**Comment by bpasero** [comment]
Thanks

**Comment by unic0rn** [comment]
long story short, the issue is still there. it seems related to the cursor and it works kinda differently than the last time i've tested, it also looks like it's not always reproducible - but i would say it is in most cases.

long story: i've tested 1.30.2, which i'm using right now, 1.31.1 (again) and 1.32.0-exploration you've provided. all versions were tested with the same exact config, extensions and workspace - everything identical. exploration was additionally tested clean - default install. each test was done several times.

now, when i've submitted this issue, 1.31.1 locked the gpu clock on 655mhz everytime it was launched, for as long as it was launched. perhaps it was related to the workspace state back then - not sure, that's my guess. testing back then was with vscode maximized and hwmon sitting on top - so vscode was not the active window.

today the gpu was sitting on 655mhz only with vscode window being active - checked with vscode resized and hwmon sitting next to it. both 1.31.1 and 1.32.0-exploration exhibit this behaviour. as long as the vscode window is active and there's a blinking cursor - gpu sits at 655mhz. i didn't check with cursor not blinking, and i didn't check without cursor in 1.31.1 - but in 1.32.0-exploration with clean config i've noticed that on welcome screen the problem doesn't manifest - gpu sits at 351mhz. as soon as i've clicked new file though - boom, 655mhz, so it has something to do with screen repaint/compositing triggered by cursor. gpu usage was basically 0, according to hwmon at least, but it was still clocked higher. once though 1.32.0-exploration, with my own config, workspace and so on, calmed down and gpu started idling at 351mhz despite having a file opened and cursor visible. it was a random occurence, i've restarted it and couldn't reproduce it since then.

in case of 1.30.2, the gpu clock doesn't even twitch. there's an occasional spike sometimes when alt-tabing, but that's normal, it's for a split second and usually doesn't even happen. with 1.31.1 and 1.32.0-exploration, when gpu sits at 655mhz when vscode window is active, it needs a second or two to go back to 351mhz after tabbing out of vscode.

**Comment by unic0rn** [comment]
i've decided to give the latest version a try - no go, same bug - and then i've found a workaround.

i've copied libEGL.dll and libGLESv2.dll from 1.30.2 to 1.32.3 and it works, GPU clock stays at 351mhz, blinking cursor and all.

bottom line, blame ANGLE.

**Comment by unic0rn** [comment]
related bug report: https://bugs.chromium.org/p/angleproject/issues/detail?id=3409

**Comment by unic0rn** [comment]
upgraded to 1.33.1 (no idea why the manual download pointed to 1.32.3 earlier - probably because i didn't restart vscode in a while), bug still present, libEGL/libGLESv2 workaround works.

**Comment by unic0rn** [comment]
CURRENT WORKAROUND: Code.exe --use-angle=gl - don't replace ANGLE dlls with older versions, there's no need, it only causes startup trouble with 1.34.0. everything below is mostly about running vscode with default settings (so via dx11 ANGLE backend).

TL;DR: check --show-paint-rects with solid block cursor for example (vim extension, no cursor blinking) - there should be no repaint, yet there is, in the top left corner of the cursor. after switching between buffers, that repaint stops, but as soon as you'll move the cursor, it comes back.

1.34.0 - not sure what happened, somehow it seems worse after update, even after reverting to 1.33.1 - i thought system reboot had some influence, cleared the shader cache, no change. and it's worse because now even with libEGL/libGLESv2 from 1.30.2 the GPU clock tends to lock on 655mhz easily. alt-tabbing out of vscode brings it back to base frequency, as well as - see above - switching between buffers back and forth. cursor set to solid.

when using something that updates the status line every second or so though, like a TODO extension, the GPU clock seems to be stuck even when tabbed out of vscode.

using 1.34.0 with default ANGLE now - not sure if there's a difference, under 1.34.0 at least, since clock seems to get locked either way, but using older ANGLE causes vscode to start slowly ("lemme hang for a sec" - black window - started). i figured a workaround though, as mentioned earlier - --use-angle=gl causes the GPU clock to stick mostly to base frequency as it should, power usage according to hwmon during heavy scrolling is on the level of dx11 backend (dx9 backend seems to draw a bit more power, while also eliminating the GPU clock problem - but hwmon isn't really a good measure of power draw i guess).

in the end, problem seems to be with dx11 backend in ANGLE, although i've got a feeling that cursor really shouldn't get repainted when not blinking, and that's probably something specific to vscode.

**Comment by unic0rn** [comment]
related problem: Code.exe -s reports complete BS in regard to actual gpu status, see https://bugs.chromium.org/p/angleproject/issues/detail?id=3409 - simplest way to check is enabling d3d9 ANGLE backend, gpu process crashes so it's not on the process list, yet vscode reports everything as working. it could be useful to have some shortcut to open new window with chrome://gpu output in the command palette.
